File: Tron/TronGameServer/Map/MapUtilities.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Tron.Utilities;

namespace Tron.GameServer
{
    public class MapUtilities : IDisposable
    {        
        private Size _mapSize;
        private Size _halfMapSize;
        private Size _floorTileSize;
        private Size _dimensions;

        public MapUtilities(Size mapSize, Size floorTileSize)
        {
            _mapSize = mapSize;
            _halfMapSize = _mapSize * .5;
            _floorTileSize = floorTileSize;
            _dimensions = new Size(_mapSize.Width / _floorTileSize.Width, _mapSize.Height / _floorTileSize.Height);
        }

        public MapLocation ToMapLocation(Vector3 position)
        {
            return new MapLocation(Math.Abs((position.z + _halfMapSize.Height) / _floorTileSize.Height), Math.Abs((position.x + _halfMapSize.Width) / _floorTileSize.Width));
        }

        public Vector3 ToPosition(MapLocation position, double y)
        {
            return new Vector3(position.Column * _floorTileSize.Width - _halfMapSize.Width, y, position.Row * _floorTileSize.Height - _halfMapSize.Height);
        }

        public bool OutOfBounds(MapLocation location)
        {
            return RowOutOfBounds(location.Row) || ColumnOutOfBounds(location.Column);
        }

        public bool OutOfBounds(Vector3 position)
        {
            return XOutOfBounds(position.x) || ZOutOfBounds(position.z);
        }

        public bool RowOutOfBounds(int row)
        {
            return row < 0 || row >= _dimensions.Height;
        }

        public bool XOutOfBounds(double x)
        {
            return x < -_halfMapSize.Width || x > _halfMapSize.Width;
        }

        public bool ZOutOfBounds(double z)
        {
            return z < -_halfMapSize.Height || z > _halfMapSize.Height;
        }

        public bool ColumnOutOfBounds(int column)
        {
            return column < 0 || column >= _dimensions.Width;
        }

        public MapLocation GetCycleMapLocation(Cycle cycle)
        {
            // Normalize to the quadrant in which the cycle lies
            MapLocation quadrant = ToMapLocation(cycle.MovementController.GetLinePosition(cycle.MovementController.Position));

            return quadrant;
        }

        public void Dispose()
        {
            _mapSize = null;
            _halfMapSize = null;
            _floorTileSize = null;
            _dimensions = null;
        }
    }
}

File: CycleR/CycleR.Game.Client/Client/Space/MapUtilities.ts
/// <reference path="../Interfaces/ThreeJS/Three.d.ts" />
/// <reference path="../Utilities/Size.ts" />
/// <reference path="../Cycles/CycleMovementController.ts" />
/// <reference path="MapLocation.ts" />

class MapUtilities {
    private _halfMapSize: Size;
    private _dimensions: Size;

    constructor(private _mapSize: Size, private _floorTileSize: Size) {
        this._halfMapSize = _mapSize.Half();
        this._dimensions = new Size(this._mapSize.Width / this._floorTileSize.Width, this._mapSize.Height / this._floorTileSize.Height);
    }

    public ToMapLocationFromController(cycleMovementController: CycleMovementController): MapLocation {
        var value = cycleMovementController.Velocity.singleValue();
        var positiveVelocity = true;

        if (value < 0) {
            positiveVelocity = false;
        }
        // if value is 0 we just assume positvie velocity because there is no movement going on

        return this.ToMapLocation(cycleMovementController.Context.position, positiveVelocity);
    }

    public ToMapLocation(position: IVector3, positiveVelocity?: bool): MapLocation {
        if (positiveVelocity) {
            return new MapLocation(Math.floor((position.z + this._halfMapSize.Height) / this._floorTileSize.Height), Math.floor((position.x + this._halfMapSize.Width) / this._floorTileSize.Width));
        }
        else {
            return new MapLocation(Math.ceil((position.z + this._halfMapSize.Height) / this._floorTileSize.Height), Math.ceil((position.x + this._halfMapSize.Width) / this._floorTileSize.Width));
        }
    }

    public ToPosition(location: MapLocation, y: number): IVector3 {
        return new THREE.Vector3(location.Column * this._floorTileSize.Width - this._halfMapSize.Width, y, location.Row * this._floorTileSize.Height - this._halfMapSize.Height);
    }

    public PositionOutOfBounds(position: IVector3): bool {
        return this.XOutOfBounds(position.x) || this.ZOutOfBounds(position.z);
    }

    public LocationOutOfBounds(location: MapLocation): bool {
        return this.RowOutOfBounds(location.Row) || this.ColumnOutOfBounds(location.Column);
    }

    public RowOutOfBounds(row: number): bool {
        return row < 0 || row >= this._dimensions.Height;
    }

    public ColumnOutOfBounds(column: number): bool {
        return column < 0 || column >= this._dimensions.Width;
    }

    public XOutOfBounds(x: number): bool {
        return x < -this._halfMapSize.Width || x > this._halfMapSize.Width;
    }

    public ZOutOfBounds(z: number): bool {
        return z < -this._halfMapSize.Height || z > this._halfMapSize.Height;
    }
}
